### ISSUE TITLE
remote fish shell initialization issues

### ISSUE THREAD

**Issue by jhheider** [warpdotdev/Warp#2454]
### Local Shell and Version

5.8.1

### Remote Shell and Version

3.3.1 (fish)

### Operating System and Version

macOS 12

### Warp Version

v0.2023.01.03.08.03.stable_01

### I have tried commenting out my system & user rcfiles and warp still will not bootstrap

- [X] Yes

### Screenshots

Similar to https://github.com/warpdotdev/Warp/issues/2201; fish attempts to send bash/zsh shell initializations, causing fish to error, preventing login.
```sh
ssh nuc.jacobsdomain.arpa
fish: Unsupported use of '='. In fish, please use 'set _user "$(whoami 2>/dev/null || echo $USER)"'.

Connection to nuc.jacobsdomain.arpa closed.
ssh nuc.jacobsdomain.arpa -vvvv
OpenSSH_9.0p1, LibreSSL 3.3.6
[...]
debug3: Ignored env NVM_DIR
debug3: Ignored env NVM_CD_FLAGS
debug3: Ignored env NVM_BIN
debug3: Ignored env NVM_INC
debug3: Ignored env PNPM_HOME
debug3: Ignored env CONDA_CHANGEPS1
debug3: Ignored env WARP_PS1
debug3: Ignored env _
debug1: Sending command:
                         export TERM_PROGRAM='WarpTerminal'

                                                           _user="$(whoami 2>/dev/null || echo $USER)"
                                                                                                      _hostname="$(hostname 2>/dev/null || uname -n)"
                                                                                                                                                     hook=$(printf "{\\"hook\\": \\"SSH\\", \\"value\\": {\\"socket_path\\": \\"~/.ssh/16730788505631\\", \\"user\\": \\"%s\\", \\"machine\\": \\"%s\\", \\"remote_shell\\": \\"%s\\"}}" "$_user" "$_hostname" "${SHELL##*/}" | command od -An -v -tx1 | command tr -d " \\n")
                                                                            printf '\033P$d%s\234' $hook

                                                                                                        if test "${SHELL##*/}" != "bash" -a "${SHELL##*/}" != "zsh"; then
                                                                                                                                                                           exec $SHELL
             fi

               case ${SHELL##*/} in
                                     bash) exec -a bash bash --rcfile <(echo 'set -o vi; set +o vi; HISTCONTROL=ignorespace; printf '\\''\033P$d7b22686f6f6b223a2022496e69745368656c6c222c202276616c7565223a207b227368656c6c223a202262617368227d7d\234'\\''')
                                                                                           ;;
                                                                                               zsh) WARP_TMP_DIR=$(mktemp -d warptmp.XXXXXX)
                                                                                                                                            local ZSH_ENV_SCRIPT='756e7365746f7074205a4c453b20756e736574205243533b20756e73657420474c4f42414c5f5243533b207072696e746620275c7831
[...]
fish: Unsupported use of '='. In fish, please use 'set _user "$(whoami 2>/dev/null || echo $USER)"'.
[...]
Connection to nuc.jacobsdomain.arpa closed.
Transferred: sent 5200, received 4720 bytes, in 0.3 seconds
Bytes per second: sent 16654.9, received 15117.5
debug1: Exit status 123
```

Can be worked around, ironically, by `ssh nuc.jacobsdomain.arpa fish`.

### Include shell xtrace output

```sh
WARP_DEBUG_MODE=1 ssh nuc.jacobsdomain.arpa
fish: Unsupported use of '='. In fish, please use 'set _user "$(whoami 2>/dev/null || echo $USER)"'.

Connection to nuc.jacobsdomain.arpa closed.
```

### Does this block you from using Warp daily?

No

### Warp Internal (ignore) - linear-label:e7dfaa84-5fdb-4a00-b754-d8912da923fa

None

### Warp Internal (ignore): linear-label:b8107fdf-ba31-488d-b103-d271c89cac3e

None

**Comment by dannyneira** [comment]
Thanks for taking the time to submit it. This is less than ideal but as a workaround, you can just disable the SSH Wrapper in `Settings > Features`. We're working on some improvements to ssh and fish and will post any updates here.

**Comment by vorporeal** [comment]
Hey @jhheider, thanks for the report!  I have a (not-production-ready) fix that gets `fish` working as a fully-featured remote shell (as long as you're running fish >=2.7.0, which is 5 years old).  I'm hoping to get it in a launchable state within the next month.

In terms of your workaround - it works because we bypass the SSH wrapper when you specify a command to run on the remote (we assume you're not starting an interactive session, so there's no reason to set up the shell with Warp's full functionality).  You can also explicitly invoke `/usr/bin/ssh` to bypass the wrapper, which is implemented by defining a function in the shell called `ssh` (to handle the typical method by which users invoke the binary).

I'm working more broadly on figuring out ways to have our SSH wrapper be more robust to things like non-Bourne shells, but that'll take longer (as it requires more thought and design).

**Comment by beejiujitsu** [comment]
Another workaround: run `fish` in the terminal prior to running `ssh` to the remote fish-shelled host. I assume this also bypasses the ssh wrapper.

**Comment by vorporeal** [comment]
Update here - I submitted a change that will get fish working as a remote shell, but without any of the fancy Warp-ification that you get with bash and zsh as remote shells.  I also have written code that adds full support for fish as a remote shell, but need to do a bit more testing and improve performance a bit before landing that functionality.

tl;dr: things should get much better within the next 2-3 weeks. :)

**Comment by dannyneira** [comment]
Hey Folks, There is a known issue with older versions of fish ( before 3.4.0 ) and Warp. Once fish is updated (the latest is 3.6.1 ), please comment if you're still seeing the same slow startup issues.

**Comment by kdrag0n** [comment]
@dannyneira I'm still getting this fatal error (not slow startup) with fish 3.6.1 on the remote host.

```
fish: Unsupported use of '='. In fish, please use 'set _user "$(command -v whoami >/dev/null 2>&1 && command whoami 2>/dev/null || echo $USER)"'.
_user="$(command -v whoami >/dev/null 2>&1 && command whoami 2>/dev/null || echo $USER)"
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
Connection to 127.0.0.1 closed.
```

**Comment by olly-writes-code** [comment]
Also getting this error

**Comment by sponge1220** [comment]
Are there any updates? Also getting this error.

**Comment by eth0net** [comment]
I'm seeing similar issues with remote fish, I did find that if I run ssh through a fish subshell without warpifying it works. 

https://app.warp.dev/block/9KFKKUjSadgktOTH1Q1VJc

https://app.warp.dev/block/b4JvWrs9HsS3hTmFuVQG1N

**Comment by hansstimer** [comment]
Looking forward to getting this fixed. For now I'm switching to zsh, but I would definitely like to be back fish.

**Comment by eth0net** [comment]
I ended up running into a bunch of similar issues and had to give it up

Not sure if it's related, but I suspect it's along the same lines, JetBrains running on Windows fails to load git from inside WSL due to a whole bunch of characters returned ahead of stdout. I reported it to them to see if they could fix it, but I suspect it may be a larger compatibility issue between fish and other systems calling into it.

For now at least, I have to stick with zsh...

**Comment by alexclst** [comment]
I see the same issue. For now I've just turned off the Warp SSH wrapper, but looking forward to being able to turn it back on once fish is supported as a remote shell.

**Comment by TangMonk** [comment]
same problem:

<img width="1014" alt="image" src="https://github.com/user-attachments/assets/90664032-44c1-49db-9d63-b661ca0f6421">

**Comment by TangMonk** [comment]
Still not fixed?

**Comment by gullitmiranda** [comment]
I don't think so. 

But I found a workaround, you can pass the shell command you want to use and the error is suppressed.

```shell
~ $ ssh old-mac
fish: Unsupported use of '='. In fish, please use 'set hook $(printf "{\"hook\": \"SSH\", \"value\": {\"socket_path\": \"~/.ssh/29272\", \"remote_shell\": \"%s\"}}" "${SHELL##*/}" | command od -An -v -tx1 | command tr -d " \n")'.
hook=$(printf "{\"hook\": \"SSH\", \"value\": {\"socket_path\": \"~/.ssh/29272\", \"remote_shell\": \"%s\"}}" "${SHELL##*/}" | command od -An -v -tx1 | command tr -d " \n")
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
Connection to macbook-pro-2.local closed.


~ $ ssh old-mac fish
direnv: loading ~/.envrc
direnv: export +DOCKER_BUILDKIT +OP_ACCOUNT +SSH_AUTH_SOCK
```

**Comment by edmcman** [comment]
Still getting this error.  👎

**Comment by edmcman** [comment]
This workaround works fine for me: https://docs.warp.dev/features/warpify/ssh-legacy#ssh-wrapper-fails

Basically you use the Warpify mechanism, instead of the SSH wrapper.  And the remote shell is still Warpified!
